google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java

```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
+     *     Redirects</a>
+     * @see <a
+     *     href="https://cloud.google.com/load-balancing/docs/https/adding-backend-buckets-to-load-balancers">
+     *     GCLB Redirects</a>
+     */
+    public static V4PostPolicyOption withBucketBoundHostname(String bucketBoundHostname) {
+      return withBucketBoundHostname(bucketBoundHostname, UriScheme.HTTP);
+    }
+
+    /**
+     * Use a bucket-bound hostname, which replaces the storage.googleapis.com host with the name of
+     * a CNAME bucket, e.g. a bucket named 'gcs-subdomain.my.domain.tld', or a Google Cloud Load
+     * Balancer which routes to a bucket you own, e.g. 'my-load-balancer-domain.tld'. Note that this
+     * cannot be used alongside {@code withVirtualHostedStyle()} or {@code withPathStyle()}. The
+     * bucket name itself should not include the URI scheme (http or https), so it is specified via
+     * a local enum.
+     *
+     * @see <a href="https://cloud.google.com/storage/docs/request-endpoints#cname">CNAME
+     *     Redirects</a>
+     * @see <a
+     *     href="https://cloud.google.com/load-balancing/docs/https/adding-backend-buckets-to-load-balancers">
+     *     GCLB Redirects</a>
+     */
+    public static V4PostPolicyOption withBucketBoundHostname(
+        String bucketBoundHostname, UriScheme uriScheme) {
+      return new V4PostPolicyOption(
+          V4PostPolicyOption.Option.BUCKET_BOUND_HOST_NAME,
+          uriScheme.getScheme() + "://" + bucketBoundHostname);
+    }
+  }
+
+  class V4PostPolicy {
+    private String url;
+    private Map<String, String> fields;
+
+    private V4PostPolicy(String url, Map<String, String> fields) {
+      this.url = url;
+      this.fields = fields;
+    }
+
+    public static V4PostPolicy of(String url, Map<String, String> fields) {
+      return new V4PostPolicy(url, fields);
+    }
+
+    public String getUrl() {
+      return url;
+    }
+
+    public Map<String, String> getFields() {
+      return fields;
+    }
+  }
+
+  class V4PostFields {
+    private Map<String, String> fieldsMap;
+    private String acl;
+    private String cacheControl;
+    private String contentDisposition;
+    private String contentEncoding;
+    private String contentLength;
+    private String contentType;
+    private String expires;
+    private String successActionRedirect;
+    private String successActionStatus;
+
+    private V4PostFields(Builder builder) {
+      this.acl = builder.acl;
+      this.cacheControl = builder.cacheControl;
+      this.contentDisposition = builder.contentDisposition;
+      this.contentEncoding = builder.contentEncoding;
+      this.contentLength = builder.contentLength;
+      this.contentType = builder.contentType;
+      this.expires = builder.expires;
+      this.successActionRedirect = builder.successActionRedirect;
+      this.successActionStatus = builder.successActionStatus;
+
+      this.fieldsMap = new HashMap<>();
```

Comment:
There's a constructor and builder and they're both setting the fieldsMap to a new mao. This is at least hard to follow and might indeed be a big. Making fieldsMap final and initializing it on the field declaration instead of in the constructor would eliminate this possibility.